### PR TITLE
make reuse of batchnorm objects possible

### PR DESCRIPTION
**Pull request by jrock08**
@tomlepaine I think you might dislike this, but I want to add reuse to the batch normalization, similar to the reuse in the simple layers.  It's not super pretty, but it does work, and might be the best/only way.

### CODE REVIEW

**Comment by tomlepaine** [comment]
Hey @jrock08, can you describe a little bit about what you are trying to do/what the expected use case is?

Also this makes the `batch_normalize` function considerable harder to parse. Can we simplify it somehow? Helper function or something. The reuse feature takes up the majority of the layer function now 😭 .

**Comment by jrock08** [comment]
So the basic idea is that we want to be able to reuse batch norm layers, similar to how we can reuse fully connected/conv layers.

Unfortunately, this is complicated.  In an ideal world, we could just reuse vars inside the batch norm object, but due to the moving average that's not possible.  Even if we share the (local) mean and variance, we have to also share the moving average object, which won't happen easily.  If we want, we could extract the moving average variable and then make the additional batch norm objects untrainable, but that's probably not ideal.

As far as the API change is concerned, it's about as low impact as it can be.  Alternatives include adding another create_batch_norm function which returns a batch norm object as well as the layer result, and then allow the user to keep track of the objects, but that seemed less magical, and really hurts simple reuse.  As it is, this hides the fact that batch norm is actually doing something weird in the background, which I really like.

As far as the code goes, yes, it's a decent number more lines, but it's mostly just that this is verbose code.  It creates a table of batch norm objects in a group and then adds to that table or retrieves from that table as necessary. That's probably a somewhat weird way to use groups, but I fail to see a much better way to do this.

**Comment by tomlepaine** [comment]
Let's break this down a little bit.

When you say reuse the batch norm layer do you want to:
- [ ] Reuse the offset/scale.
- [ ] Reuse the mean/variance (this seems weird because the mean and variance should be calculated on its inputs).
- [ ] Reuse the exponential moving average mean/variance. (Why should one layer's mean and variance modify another layers mean and variance?)

I guess I don't get the use case. Do you want the output of layer X to be normalized by the parameters of layer Y? Is that they idea? If so it seems an unusual enough use case, that I don't think it should be part of the regular batch_normalize layer. Maybe it belongs in another layer that is specifically for this use case?

**Comment by jrock08** [comment]
So, lets say I have an adversarial network.  I need to run that network on the output from a generator (for all of the parameters of the network) when I train the generator.  The easiest way to do that seems to be to duplicate the network with input equal to output of the generator.

**Comment by BigRedT** [comment]
@jrock08: why is this batch_normalization (probably @tomlepaine or @mbz 's) code so much simpler than the one in this PR?
https://github.com/mbz/UIUCyberdemon/blob/master/batch_normalizer.py

**Comment by jrock08** [comment]
That implementation won't work for running single images through at test time. Something that is necessary for image to image generation code, since at test time, "batch" = 1

**Comment by tomlepaine** [comment]
Why doesn't this work for a single image?

**Comment by BigRedT** [comment]
@tomlepaine what does tf.nn.moments given for variance when there's a single image?

**Comment by kevjshih** [comment]
The mean and variance updating should be moved into the `if train:` block for this code, after which I think it'll be pretty similar to what Jason pushed. If your test comes from a very different distribution as your training set (ie. train on positive and negatives and test on only negatives), the shifted mean and variance during test time will break your model because the rest of your network isn't changing but the  normalization parameters between layers are. In the given example, it's conceivable that the negatives will start looking like positives after a while because of the shift in mean.

**Comment by jrock08** [comment]
Oh, I actually missed that you had adjusted that code.  That is better, but the way the train and test works, you won't be able to actually use it for test because if you set it to test the ema will never get anything except default values.  You'd have to reuse the ema from somewhere.  You can get the reuse of the ema by replacing the` train=True` to `train = tf.placeholder('train_phase')`, and then handling it with control_flow_ops should.  But still won't allow for reuse_var in a different location (which is the point of this pr).

Reuse would require recovering the ema from somewhere, and once you do that, you might as well wrap the ema into a batch norm object.

```
control_flow_ops.cond(phase_train,
            lambda: bn.normalize(input, True),
            lambda: bn.normalize(input, False))
```

Creation (or not) of a batch norm object to wrap all this up is a personal preference thing.  I think that creating a batch norm object, and then wrapping it's construction inside functions makes sense, you could obviously unwrap the object and write them directly in the functions.

**Comment by tomlepaine** [comment]
Oh... maybe I am confused. Do you actually make new exponential moving
averages at test time?

This code basically has an exponential moving during training. And uses
that during test.
When it checkpoints it saves `average_mean`, and `average_variance`, and
then at test time, it restores `average_mean`, and `average_variance` and
uses them in `tf.batch_normalization`.

That was my goal. Is this wrong? Your resulting mean is fixed (so
predictions are deterministic), is made up of many images (so it is scaled
well), and it doesn't depend on the number of images.

**Comment by tomlepaine** [comment]
Note: As I understand it `average_mean` and `average_variance` are only
updated when you call `ema_op` so it doesn't technically matter where they
are defined, they will not be updated.

If you think this will have problems, we don't have to use it, but I think
it addresses the problems the batch norm paper mentioned with using single
batch mean and variances during testing.

**Comment by kevjshih** [comment]
My bad, I didn't realize the updates wouldn't happen unless called. Clearly still getting used to this graph-based programming.

**Comment by jrock08** [comment]
0. This would be a better discussion with another pull request/branch suggesting the changes to the other piece of code.

1. Not totally sure if that ema update works as written due to scoping of the mean and variance.  Would want to see tests that verify.

2. This code still doesn't address reusing batch norm layers, and reuse would require storing out the EMAs and then reusing them inside this.  At that point, it makes sense to wrap the EMA with a batch norm object and then store the batch norm object so it's obvious what the EMA is averaging over (and how to update it). 

When I say reuse batch norm layers I mean, I have some tensor A and some tensor B, and would like to train them such that they go through the same normalization.  E.G. siamese network: A and B get fed to the same network which shares all parameters including batch norms.  Adversarial Network: A is training data for the discriminator, and B is the generator that needs to learn to fool the discriminator which includes batch norms.

**Comment by jrock08** [comment]
I'm going to close this.  With discussion offline with @BigRedT, I think the current implementation in #28 captures what we actually want.  Reuse of the EMA doesn't make much sense.
